### PR TITLE
Added writable snapshot in info module

### PR DESCRIPTION
**Pull request by Kritika-Bhateja-03**
# Description
A few sentences describing the overall goals of the pull request's commits.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
| |

# Checklist:

- [ ] I have performed a self-review of my own code to ensure there are no formatting, pep8, linting, or security issues
- [ ] I have performed Ansible Sanity test using --docker default
- [ ] I have verified that new and existing unit tests pass locally with my changes
- [ ] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: plugins/module_utils/storage/dell/shared_library/writable_snapshot.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Copyright: (c) 2024, Dell Technologies
```
Comment (Bhavneet-Sharma) [inline plugins/module_utils/storage/dell/shared_library/writable_snapshot.py:1]: I guess this file is not required; we can add these methods into existing snapshot file as it comes under snapshot_api as well.

File: playbooks/modules/info.yml [hunk 2]
```diff
@@ -395,3 +395,50 @@
...
+        api_user: "{{ api_user }}"
+        api_password: "{{ api_password }}"
+        gather_subset:
+          - writable_snapshot
+        query_parameters:
+          writable_snapshot:
+            dir: ASC
+            limit: 1
+
+    - name: To sort the writable snapshot in ascending order.
```
Comment (meenakshidembi691) [inline playbooks/modules/info.yml:433]: Please update the name of the task query parameter is sort

File: plugins/modules/info.py [hunk 3]
```diff
@@ -184,13 +185,38 @@
     description:
     - Contains dictionary of query parameters for specific I(gather_subset).
     - Applicable to C(alert_rules), C(event_group), C(event_channels) and C(filesystem).
+    - If C(writable_snapshot) is passed as I(gather_subset), if I(wspath) is given,
```
Comment (sachin-apa) [inline plugins/modules/info.py:188]: should we keep the sub options, for event channels and rules sub options were not added.
@Bhavneet-Sharma @meenakshidembi691 FYI , please comment.

File: playbooks/modules/info.yml (hunk 2 above)
Comment (Kritika-Bhateja-03) [reply playbooks/modules/info.yml]: done

File: playbooks/modules/info.yml [hunk 4]
```diff
@@ -395,3 +395,50 @@
...
+    - name: Get all writable snapshots from PowerScale cluster
+      dellemc.powerscale.info:
+        onefs_host: "{{ onefs_host }}"
+        verify_ssl: "{{ verify_ssl }}"
+        api_user: "{{ api_user }}"
+        api_password: "{{ api_password }}"
+        gather_subset:
+          - writable_snapshot
+
+    - name: To get the specific writable snapshot.
```
[the remark this answers is not in the document]
Comment (Kritika-Bhateja-03) [reply playbooks/modules/info.yml]: done

File: plugins/module_utils/storage/dell/shared_library/writable_snapshot.py (hunk 1 above)
Comment (Kritika-Bhateja-03) [reply plugins/module_utils/storage/dell/shared_library/writable_snapshot.py]: done

File: docs/modules/info.rst [hunk 5]
```diff
@@ -782,6 +810,53 @@ Examples
...
+    - name: Get all writable snapshots from PowerScale cluster
+      dellemc.powerscale.info:
+        onefs_host: "{{ onefs_host }}"
+        verify_ssl: "{{ verify_ssl }}"
+        api_user: "{{ api_user }}"
+        api_password: "{{ api_password }}"
+        gather_subset:
+          - writable_snapshot
+
+    - name: To get the specific writable snapshot.
```
[the remark this answers is not in the document]
Comment (Kritika-Bhateja-03) [reply docs/modules/info.rst:822]: done

**Review by sachin-apa — APPROVED** [review]

**Review by sachin-apa — DISMISSED** [review]
Validation Completed !

File: plugins/module_utils/storage/dell/shared_library/snapshot.py [hunk 6]
```diff
@@ -50,3 +50,47 @@ def get_filesystem_snapshots(self, effective_path):
...
+    def get_particular_writable_snapshot(self, wspath):
+        try:
+            return self.snapshot_api.get_snapshot_writable_wspath(
+                snapshot_writable_wspath=wspath
+            ).to_dict().get("writable")
+        except utils.ApiException as e:
+            return {}
+        except Exception as e:
+            error_msg = utils.determine_error(error_obj=e)
+            error_message = 'Failed to get writeable snapshots ' \
```
Comment (trisha-dell) [inline plugins/module_utils/storage/dell/shared_library/snapshot.py:93]: As this is for a particular snapshot change snapshots to snapshot.

File: plugins/modules/info.py [hunk 7]
```diff
@@ -183,14 +184,22 @@
...
+      all other query parameters inside I(writable_snapshot) will be ignored.
+    - To view the list of supported query parameters for C(writable_snapshot),
+      refer Query Parameters section from
+      U(https://developer.dell.com/apis/4088/versions/9.5.0/9.5.0.0_ISLANDER_OAS2.json/
+        paths/~1platform~114~1snapshot~1writable/get)
     type: dict
     version_added: '3.2.0'
 notes:
 - The parameters I(access_zone) and I(include_all_access_zones) are mutually exclusive.
 - Listing of SyncIQ target cluster certificates is not supported by isi_sdk_8_1_1 version.
```
Comment (trisha-dell) [inline plugins/modules/info.py:199]: Shall we remove this line as we do not support this version anymore?

File: plugins/modules/info.py [hunk 8]
```diff
@@ -709,6 +718,53 @@
...
+    api_user: "{{ api_user }}"
+    api_password: "{{ api_password }}"
+    gather_subset:
+      - writable_snapshot
+    query_parameters:
+      writable_snapshot:
+        dir: ASC
+        limit: 1
+
+- name: To sort the writable snapshot in ascending order.
```
Comment (trisha-dell) [inline plugins/modules/info.py:756]: rename the task.

File: plugins/modules/info.py [hunk 9]
```diff
@@ -183,14 +184,22 @@
...
     description:
     - Contains dictionary of query parameters for specific I(gather_subset).
-    - Applicable to C(alert_rules), C(event_group), C(event_channels) and C(filesystem).
+    - Applicable to C(alert_rules), C(event_group), C(event_channels), C(filesystem)
+      and C(writable_snapshot).
+    - If C(writable_snapshot) is passed as I(gather_subset), if I(wspath) is given,
+      all other query parameters inside I(writable_snapshot) will be ignored.
+    - To view the list of supported query parameters for C(writable_snapshot),
+      refer Query Parameters section from
+      U(https://developer.dell.com/apis/4088/versions/9.5.0/9.5.0.0_ISLANDER_OAS2.json/
```
Comment (trisha-dell) [inline plugins/modules/info.py:193]: can we provide the link to the latest version?

Comment (trisha-dell) [reply plugins/modules/info.py:193]: U before the link seems to be a typo

File: plugins/module_utils/storage/dell/shared_library/snapshot.py [hunk 10]
```diff
@@ -50,3 +50,47 @@ def get_filesystem_snapshots(self, effective_path):
...
+            writable_snapshots.extend(snapshot_list['writable'])
+            return writable_snapshots
+        except Exception as e:
+            error_msg = utils.determine_error(error_obj=e)
+            error_message = 'Failed to get writeable snapshots ' \
+                            'due to error {0}'.format((str(error_msg)))
+            LOG.error(error_message)
+            self.module.fail_json(msg=error_message)
+
+    def get_particular_writable_snapshot(self, wspath):
```
[the remark this answers is not in the document]
Comment (Kritika-Bhateja-03) [reply plugins/module_utils/storage/dell/shared_library/snapshot.py]: done

File: plugins/module_utils/storage/dell/shared_library/snapshot.py (hunk 6 above)
Comment (Kritika-Bhateja-03) [reply plugins/module_utils/storage/dell/shared_library/snapshot.py]: done

File: plugins/modules/info.py (hunk 8 above)
Comment (Kritika-Bhateja-03) [reply plugins/modules/info.py]: done

**Review by sachin-apa — DISMISSED** [review]

**Review by felixs88 — APPROVED** [review]

**Review by felixs88 — APPROVED** [review]

**Review by sachin-apa — DISMISSED** [review]

**Review by meenakshidembi691 — APPROVED** [review]

**Review by trisha-dell — DISMISSED** [review]

File: tests/unit/plugins/module_utils/mock_info_api.py [hunk 11]
```diff
@@ -1636,6 +1636,41 @@ def get_filesystem_list(response_type):
         else:
             return "Failed to get details of Filesystem"
 
+    @staticmethod
+    def get_writable_snapshot_list(response_type):
+        if response_type == 'api':
+            return {
```
Comment (Bhavneet-Sharma) [inline tests/unit/plugins/module_utils/mock_info_api.py:1642]: if data is same then, I guess if-else block is not required

**Review by Bhavneet-Sharma — COMMENTED** [review]
I have a comment

File: tests/unit/plugins/module_utils/mock_info_api.py (hunk 11 above)
Comment (Kritika-Bhateja-03) [reply tests/unit/plugins/module_utils/mock_info_api.py:1642]: We have a difference of a key "writable" in first case.

**Review by Bhavneet-Sharma — APPROVED** [review]

**Review by Bhavneet-Sharma — APPROVED** [review]

**Review by trisha-dell — APPROVED** [review]

**Review by sachin-apa — APPROVED** [review]

**Pull request merged**